packages/server/src/fhir/references.ts

```diff
@@ -1,105 +1,136 @@
-import { getSystemRepo } from './repo';
+import { Repository, setTypedPropertyValue } from './repo';
 
-export async function validateReferences<T extends Resource>(
-  resource: T,
-  allowedProjects?: Project['id'][]
-): Promise<void> {
-  return new FhirReferenceValidator(resource, allowedProjects).validate();
-}
+type ScanFn = (tv: TypedValue, path: string) => Promise<TypedValue | undefined>;
+
+export async function validateReferences<T extends Resource>(resource: T, repo: Repository): Promise<void> {
+  const issues: OperationOutcomeIssue[] = [];
+  await new FhirResourceScanner(resource).forEach(PropertyType.Reference, async (value, path) => {
+    const reference = value.value as Reference;
+    if (reference.reference?.split?.('/')?.length !== 2) {
+      return undefined;
+    }
 
-export class FhirReferenceValidator<T extends Resource> {
-  private readonly issues: OperationOutcomeIssue[];
-  private readonly allowedProjects?: Project['id'][];
-  private readonly root: T;
+    try {
+      await repo.readReference(reference);
+    } catch (err) {
+      issues.push(createStructureIssue(path, `Invalid reference (${normalizeErrorString(err)})`));
+    }
+    return undefined;
+  });
 
-  constructor(root: T, allowedProjects?: Project['id'][]) {
-    this.issues = [];
-    this.allowedProjects = allowedProjects;
-    this.root = root;
+  if (issues.length > 0) {
+    throw new OperationOutcomeError({
+      resourceType: 'OperationOutcome',
+      id: randomUUID(),
+      issue: issues,
+    });
   }
+}
+
+export async function replaceConditionalReferences<T extends Resource>(resource: T, repo: Repository): Promise<T> {
+  const scanner = new FhirResourceScanner(resource);
+  await scanner.forEach(PropertyType.Reference, async (value, path) => {
+    const reference = value.value as Reference;
+    if (!reference.reference?.includes?.('?')) {
+      return undefined;
+    }
 
-  async validate(): Promise<void> {
-    const resource = this.root;
-    const resourceType = resource.resourceType;
-    await this.validateObject(resourceType, toTypedValue(resource));
-
-    if (this.issues.length > 0) {
-      throw new OperationOutcomeError({
-        resourceType: 'OperationOutcome',
-        id: randomUUID(),
-        issue: this.issues,
-      });
+    const searchCriteria = parseSearchRequest(reference.reference);
+    searchCriteria.sortRules = undefined;
+    searchCriteria.count = 2;
+    const matches = await repo.searchResources(searchCriteria);
+    if (matches.length !== 1) {
+      throw new OperationOutcomeError(
+        badRequest(
+          `Conditional reference '${reference.reference}' ${matches.length ? 'matched multiple' : 'did not match any'} resources`,
+          path
+        )
+      );
     }
+
+    const resolvedReference = createReference(matches[0]);
+    return { type: PropertyType.Reference, value: resolvedReference };
+  });
+
+  return resource;
+}
+
+export class FhirResourceScanner {
```

Comment:
`ResourceCrawler` addresses a similar use case, but its method of operation would need to be extended to handle reference resolution.  At the very least, it would need to become async, which would break method contracts in `@medplum/core`

Reply:
Let's talk through the changes we'd need to make. It might be worth it to consolidate